File: src/components/counter.c
#include "idom.h"

static int count = 0;

HANDLER(on_decrement) { count--; }
HANDLER(on_increment) { count++; }

INIT {
  TEMPLATE(0);
  EL("div") {
    ATTR("class", "counter");
    EL("button") {
      ON("click", on_decrement);
      TEXT("\xe2\x88\x92");
    } END;
    EL("span") {
      ATTR("class", "count");
      SLOT;
    } END;
    EL("button") {
      ON("click", on_increment);
      TEXT("+");
    } END;
  } END;
  END_TEMPLATE;
}

EXPORT render() {
  UPDATE(0);
  value_int(count);
  COMMIT;
}

File: src/components/counter.zig
const dom = @import("dom.zig");

var count: i32 = 0;

export fn on_decrement() void { if (count > 0) count -= 1; }
export fn on_increment() void { count += 1; }

const view = dom.html(
    \\<div class='counter'>
    \\  <button on:click='on_decrement'>-</button>
    \\  <span class='count'>{}</span>
    \\  <button on:click='on_increment'>+</button>
    \\</div>
);

export fn render() u32 {
    return view.update(.{count});
}
